### PR TITLE
Simple BRNN

### PR DESCRIPTION
**Pull request by patyork**
Added a layer in the style of SimpleRNN called SimpleBRNN. This is a fully connected Bidirectional recurrent layer.

My only issue with this implementation is that the initialization looks like so:
```python
model.add(SimpleBRNN(5, output_dim=2*10, activation='sigmoid', return_sequences=True))
```
The ```output_dim=2*10``` is meant to imply that the forward recurrence has 10 outputs and the backwards recurrence has 10 outputs. These are concatenated, and the output of the SimpleBRNN layer would be 2x10=20 outputs at each time step, for each sample (batch_size, time_steps, 20).

### CODE REVIEW

**Comment by fchollet** [comment]
I'm curious as to what the background on this is: do you achieve interesting results with this architecture? Is this documented in a publication somewhere?

If this layer turns out to be useful (similarly to the usefulness of GRU and LSTM) I will merge it. Otherwise, I'd prefer to keep "demo" architectures to a minimum. We currently have SimpleRNN and SimpleDeepRNN that are demo architectures, and they're just meant to provide accessible examples of how to build RNN layers.

**Comment by patyork** [comment]
http://arxiv.org/abs/1412.5567

Baidu (semi-)recently released this paper. It uses 3 Dense layers, a BRNN layer, and a Dense layer to achieve state-of-the-art speech recognition in noisy environments. That is the impetus of this layer.

The 'noisy' aspect is pre-processing, the rest is an easily implemented Keras model :smile: if this module is added.

I will implementing a fast CTC (http://www.autonlab.org/icml_documents/camera-ready/047_Connectionist_Tempor.pdf) objective class in the coming days/weeks.

File: keras/layers/recurrent.py [hunk 1]
```diff
@@ -66,6 +67,75 @@ def output(self, train):
...
+            Output at time=t is fed back to input for time=t+1 in a forward pass
+            Output at time=t is fed back to input for time=t-1 in a backward pass
+    '''
+    def __init__(self, input_dim, output_dim,
+        init='uniform', inner_init='orthogonal', activation='sigmoid', weights=None,
+        truncate_gradient=-1,  return_sequences=False):
+        self.init = initializations.get(init)
+        self.inner_init = initializations.get(inner_init)
+        self.input_dim = input_dim
+        self.input_dim = input_dim
```
Comment (fchollet) [inline keras/layers/recurrent.py]: redundant

File: keras/layers/recurrent.py [hunk 2]
```diff
@@ -66,6 +67,75 @@ def output(self, train):
...
+        self.activation = activations.get(activation)
+        self.return_sequences = return_sequences
+        self.input = T.tensor3()
+
+        self.W_if = self.init((self.input_dim, self.output_dim/2))    # Input -> Forward
+        self.W_ff = self.init((self.output_dim/2, self.output_dim/2))   # Forward tm1 -> Forward t
+        self.W_ib = self.init((self.input_dim, self.output_dim/2))    # Input -> Backward
+        self.W_bb = self.init((self.output_dim/2, self.output_dim/2))   # Backward t -> Backward tm1
+        self.b_f = shared_zeros((self.output_dim/2))
+        self.b_b = shared_zeros((self.output_dim/2))
```
Comment (fchollet) [inline keras/layers/recurrent.py]: This divides the input into halves of size output_dim/2 that are then concatenated. The size of the output might not be output_dim anymore (eg: python(13/2) = 6; 6+6=12).

**Comment by fchollet** [comment]
I see the Baidu paper references this older work: http://www.di.ufpe.br/~fnj/RNA/bibliografia/BRNN.pdf

Could you add these references to your class description? 

I added some comments about issues to be fixed.

Ideally, could you provide an example script (to be added to the examples folder) where the BRNN layer shows reasonably good results on a dataset?

**Comment by patyork** [comment]
If you would prefer, I will rename the class BDLayer (or some variant) to reflect that this is a Bidirectional layer that is not a demonstration.

File: keras/layers/recurrent.py (hunk 2 above)
Comment (patyork) [inline keras/layers/recurrent.py]: That was my concern in the original description. The layers should, naturally, provide 1/2 the input to the next layer. A safeguard makes sense, but should not be necessary.

Comment (fchollet) [inline keras/layers/recurrent.py]: Ok, I had overlooked that. I think this should be enforced via an exception. Knowing from the start what you did wrong is better than a failure at training time due to shape incompatibilities.

File: keras/layers/recurrent.py (hunk 1 above)
Comment (patyork) [inline keras/layers/recurrent.py]: Could you elaborate a bit on this? I see the two duplicate lines as redundant (not sure how they entered the picture), but I'm not sure how much you wan to inherit, especially given SimpleRNN is a demo implementation.

File: keras/layers/recurrent.py (hunk 2 above)
Comment (patyork) [inline keras/layers/recurrent.py]: That is fair. I will add it tomorrow.

**Comment by patyork** [comment]
The problem with this is that a BRNN only really improves when CTC is applied. I have yet to find a dataset where a BRNN improves over an RNN. Perhaps the simplest example I should try is Hinton's 'binary addition' example.

I have not tested this fully - there may be situations where a BRNN retains enough information more than an RNN to see improvement over a known, or testable, baseline.

**Comment by fchollet** [comment]
If a BRNN has to be used in conjunction with CTC to be useful, then it would be very valuable to 1) add both the BRNN layer and CTC at the same time, and 2) provide an example script with a model leveraging both, so that users have a reference point of what to do with BRNN and CTC, and what kind of performance they should expect.

Examples are exponentially more useful the more gotchas there are with the usage of a module. 

Anyway, this module seems like a great addition, I'm looking forward to see what you do with it : )

**Pull request closed without merging**